### PR TITLE
Resending to registry backend

### LINKED ISSUE #312: De medewerker die met de toepassing werkt heeft inzage in de wachtrij met formulieren die nog aan het systeem ...
...worden aangeboden. Het systeem plaatst de ingezonden formulieren in een wachtrij wanneer ze niet afgeleverd kunnen worden aan het gekoppelde backofficesysteem/zaaksysteem. Het systeem biedt op gezette tijden de wachtrij opnieuw aan (deze eis valt buiten het MVP).

Eis: Afhandeling 2.
Zie ook NvI: 62

**Tasks**

- [x] Show registration status in admin submission list view
- [x] Show last retry date in admin submission list view
- [x] Show registration backend in submission detail view (can be read only)
- [x] Add bulk action to resend submissions that have not yet been sent succesfully
- [x] Add periodic resend action to Celery Beat to resend all submissions that have not yet been sent succefully (use exponential backoff, see: https://stackoverflow.com/questions/9731435/retry-celery-tasks-with-exponential-back-off)
- [x] Stop resending automatically (celery beat) after 48 (make setting)hours (then they can still use the bulk action).

### PR DESCRIPTION
**Pull request by shea-maykinmedia**
Fixes #312 

This did feel a bit tricky to test but I tried testing all the scenarios I can think of.

**Screenshots**

Celery beat task running every minute

resend_submissions_running_every_minute

Register submissions retrying.  I set the max retries to 4 for quicker testing and verified in the admin that after the 4th time the submission status is set to failed.

testing_task_retrying

Testing the admin action.  Specifically that you can only try resending submissions in the failed state.

testing_admin_action_1

testing_admin_action_2

Testing celery beat only runs failed tasks within last 48 hours 

admin_before_celery_beat_task

testing_celery_beat_only_does_failing_within_48_hours

admin_after_celery_beat_task

### CODE REVIEW

File: src/openforms/conf/base.py [hunk 1]
```diff
@@ -515,8 +515,18 @@
         "task": "openforms.utils.tasks.send_emails",
         "schedule": config("BEAT_SEND_EMAIL_INTERVAL", default=20),  # every 20 seconds
     },
+    "resend-submissions": {
+        "task": "openforms.registrations.tasks.resend_submissions",
+        "schedule": config(
+            "BEAT_RESEND_SUBMISSIONS_INTERVAL", default=60  # every minute
```
Comment (sergei-maertens) [inline src/openforms/conf/base.py:521]: Please document this parameter in the docs

File: src/openforms/conf/base.py [hunk 2]
```diff
@@ -515,8 +515,18 @@
...
+        ),
+    },
 }
 
+CELERY_BEAT_RESEND_SUBMISSIONS_TIME_LIMIT = config(
+    "CELERY_BEAT_RESEND_SUBMISSIONS_TIME_LIMIT", default=48  # hours
```
Comment (sergei-maertens) [inline src/openforms/conf/base.py:527]: Please document this parameter in the docs

File: src/openforms/conf/base.py [hunk 3]
```diff
@@ -527,6 +537,9 @@
 # *should* have the same effect...
 CELERY_WORKER_PREFETCH_MULTIPLIER = 1
 
+# Amount of times to retry the celery task before stopping
+CELERY_MAX_RETRIES = config("CELERY_MAX_RETRIES", default=10)
```
Comment (sergei-maertens) [inline src/openforms/conf/base.py:541]: Please document this parameter in the docs

File: src/openforms/registrations/tasks.py [hunk 4]
```diff
@@ -74,7 +89,22 @@ def register_submission(submission_id: int) -> Optional[dict]:
...
+        ]
+    )
+
+
+@app.task
+def resend_submissions():
+    for submission in Submission.objects.filter(
+        registration_status=RegistrationStatuses.failed,
+        completed_on__gte=timezone.now()
+        - timedelta(hours=settings.CELERY_BEAT_RESEND_SUBMISSIONS_TIME_LIMIT),
```
Comment (sergei-maertens) [inline src/openforms/registrations/tasks.py:105]: you can move this calculation out of the loop, which will in turn make the code more readable

File: src/openforms/registrations/tasks.py [hunk 5]
```diff
@@ -54,9 +63,15 @@ def register_submission(submission_id: int) -> Optional[dict]:
...
-        result_data = {"traceback": formatted_tb}
+        if task.request.retries <= settings.CELERY_MAX_RETRIES:
+            return task.retry(
+                submission_id=submission_id,
+                countdown=3 ** task.request.retries,
+            )
+        else:
+            formatted_tb = traceback.format_exc()
+            status = RegistrationStatuses.failed
+            result_data = {"traceback": formatted_tb}
```
Comment (sergei-maertens) [inline src/openforms/registrations/tasks.py:74]: Afaik it should be possible to specify all this in the task decorator, something like:

```py
@app.task(
    bind=True,
    autoretry_For=(RegistrationFailed,),
    retry_backoff=True,  # exponential backoff
    max_retries=settings.CELERY_MAX_RETRIES
)
def register_submission(task, submission_id: int) -> Optional[dict]:

```

See the docs: https://docs.celeryproject.org/en/stable/userguide/tasks.html#Task.autoretry_for

That should in turn make this code more like this:

```py
except RegistrationFailed:
    formatted_tb = traceback.format_exc()
    submission.registration_status = RegistrationStatuses.failed
    submission.registration_result = {"traceback": formatted_tb}
    submission.save(
        update_fields=[
            "registration_status",
            "registration_result",
        ]
    )
    raise
```

Even if something is going into the retry, I'd like to see the traceback in the admin interface without having to check Flower.

File: src/openforms/conf/base.py (hunk 3 above)
Comment (sergei-maertens) [reply src/openforms/conf/base.py:541]: Additionally, this makes it seem like this is an official Celery setting, which it isn't (at least I can't find it on https://docs.celeryproject.org/en/stable/userguide/configuration.html)

If it's a custom setting (which I'm 99% certain it is), please name it appropriately. I suggest using `SUBMISSION_REGISTRATION_MAX_RETRIES`

File: src/openforms/submissions/admin.py [hunk 6]
```diff
@@ -40,8 +43,16 @@ class SubmissionAdmin(admin.ModelAdmin):
...
+    readonly_fields = [
+        "created_on",
+        "get_registration_backend",
+    ]
+    actions = ["export_csv", "export_xlsx", "resend_submissions"]
+
+    def get_registration_backend(self, obj):
+        return obj.form.registration_backend
+
+    get_registration_backend.short_description = "Registration Backend"
```
Comment (sergei-maertens) [inline src/openforms/submissions/admin.py:55]: Mark the string as translatable

File: src/openforms/submissions/models.py [hunk 7]
```diff
@@ -86,6 +86,15 @@ class Submission(models.Model):
     created_on = models.DateTimeField(_("created on"), auto_now_add=True)
     completed_on = models.DateTimeField(_("completed on"), blank=True, null=True)
     suspended_on = models.DateTimeField(_("suspended on"), blank=True, null=True)
+    last_register_date = models.DateTimeField(
+        _("last register date"),
```
Comment (sergei-maertens) [inline src/openforms/submissions/models.py:90]:
```suggestion
        _("last registration attempt date"),
```

File: src/openforms/submissions/models.py [hunk 8]
```diff
@@ -86,6 +86,15 @@ class Submission(models.Model):
...
+        blank=True,
+        null=True,
+        help_text=_(
+            "The last time the submission was registered with the backend.  "
+            "Note that this date will be updated even if the registration is not successful."
```
Comment (sergei-maertens) [inline src/openforms/submissions/models.py:95]:
```suggestion
            "The last time the submission registration was attempted with the backend.  "
            "Note that this date will be updated even if the registration is not successful."
```

**Review by sergei-maertens — CHANGES_REQUESTED** [review]
Couple beauty fixes, but overall this looks good

**Comment by shea-maykinmedia** [comment]
Retesting celery retrying with new configuration

celery_task_retrying

Result between celery retries

status_between_retrying

Documentation

documentation

**Review by sergei-maertens — APPROVED** [review]

**Comment by sergei-maertens** [comment]
Failing docs build is because opengem.nl is down

**Pull request merged**